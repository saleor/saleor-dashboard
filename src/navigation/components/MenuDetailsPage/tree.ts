import { MenuDetailsFragment } from "@saleor/graphql";

import { TreeOperation } from "../MenuItems";

export function findNode(
  tree: MenuDetailsFragment["items"],
  id: string,
): number[] {
  const foundNodeIndex = tree.findIndex(node => node.id === id);
  if (tree.length === 0) {
    return [null];
  }
  if (foundNodeIndex !== -1) {
    return [foundNodeIndex];
  }
  const nodeMap = tree.map((node, nodeIndex) => [
    nodeIndex,
    ...findNode(node.children, id),
  ]);
  return nodeMap.find(path => path[path.length - 1] !== null) || [null];
}

export function getNode(
  tree: MenuDetailsFragment["items"],
  path: number[],
): MenuDetailsFragment["items"][0] {
  if (path.length === 1) {
    return tree[path[0]];
  }
  return getNode([...tree[path[0]].children], path.slice(1));
}

function removeNode(
  tree: MenuDetailsFragment["items"],
  path: number[],
): MenuDetailsFragment["items"] {
  const removeIndex = path[0];

  if (path.length === 1) {
    return [...tree.slice(0, removeIndex), ...tree.slice(removeIndex + 1)];
  }

  const newTree = [...tree];
  newTree[removeIndex] = {
    ...tree[path[0]],
    children: removeNode(tree[path[0]].children, path.slice(1)),
  };

  return newTree;
}

interface InsertNodeInput {
  tree: MenuDetailsFragment["items"];
  path: number[];
  node: MenuDetailsFragment["items"][0];
  position: number;
}

function insertNode({
  tree,
  path,
  node,
  position,
}: InsertNodeInput): MenuDetailsFragment["items"] {
  if (path.length === 0) {
    return [...tree.slice(0, position), node, ...tree.slice(position)];
  }

  if (path[0] in tree) {
    tree[path[0]].children = insertNode({
      tree: tree[path[0]].children,
      path: path.slice(1),
      node,
      position,
    });
  }
  return tree;
}

function removeNodeAndChildren(
  tree: MenuDetailsFragment["items"],
  operation: TreeOperation,
): MenuDetailsFragment["items"] {
  const sourcePath = findNode(tree, operation.id);
  const node = getNode(tree, sourcePath);

  if (node.children) {
    const treeAfterChildrenRemoval = node.children.reduce(
      (acc, child) =>
        removeNodeAndChildren(acc, {
          id: child.id,
          type: "remove",
        }),
      tree,
    );

    return removeNode(treeAfterChildrenRemoval, sourcePath);
  }

  return removeNode(tree, sourcePath);
}

function permuteRelativeNode(
  tree: MenuDetailsFragment["items"],
  permutation: TreeOperation,
): MenuDetailsFragment["items"] {
  const sourcePath = findNode(tree, permutation.id);
  const node = getNode(tree, sourcePath);

  const hasParent = !!permutation.parentId;

  const treeAfterRemoval = removeNode(tree, sourcePath);

  const targetPath = hasParent
    ? findNode(treeAfterRemoval, permutation.parentId)
    : [];

  const position = sourcePath[sourcePath.length - 1];

  const treeAfterInsertion = insertNode({
    tree: treeAfterRemoval,
    path: targetPath,
    node,
    position: position + permutation.sortOrder,
  });

  return treeAfterInsertion;
}

function executeRelativeOperation(
  tree: MenuDetailsFragment["items"],
  operation: TreeOperation,
): MenuDetailsFragment["items"] {
  return operation.type === "move"
    ? permuteRelativeNode(tree, operation)
    : removeNodeAndChildren(tree, operation);
}

export function computeRelativeTree(
  tree: MenuDetailsFragment["items"],
  operations: TreeOperation[],
) {
  const newTree = operations.reduce(
    (acc, operation) => executeRelativeOperation(acc, operation),
    JSON.parse(JSON.stringify(tree)),
  );
  return newTree;
}
