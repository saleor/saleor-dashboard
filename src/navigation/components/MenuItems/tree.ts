import { MenuDetailsFragment } from "@saleor/graphql";
import { getPatch } from "fast-array-diff";
import { TreeItem } from "react-sortable-tree";

import { MenuItemType } from "../MenuItemDialog";

export type TreeOperationType = "move" | "remove";
export interface TreeOperation {
  id: string;
  type: TreeOperationType;
  parentId?: string;
  sortOrder?: number;
  simulatedMove?: boolean;
}

export const unknownTypeError = Error("Unknown type");

function treeToMap(tree: TreeItem[], parent: string): Record<string, string[]> {
  const childrenList = tree.map(node => node.id);
  const childrenMaps = tree.map(node => ({
    id: node.id,
    mappedNodes: treeToMap(node.children as TreeItem[], node.id),
  }));

  return {
    [parent]: childrenList,
    ...childrenMaps.reduce(
      (acc, childMap) => ({
        ...acc,
        ...childMap.mappedNodes,
      }),
      {},
    ),
  };
}

export function getItemType(
  item: MenuDetailsFragment["items"][0],
): MenuItemType {
  if (item.category) {
    return "category";
  } else if (item.collection) {
    return "collection";
  } else if (item.page) {
    return "page";
  } else if (item.url) {
    return "link";
  } else {
    throw unknownTypeError;
  }
}

export function getItemId(item: MenuDetailsFragment["items"][0]): string {
  if (item.category) {
    return item.category.id;
  } else if (item.collection) {
    return item.collection.id;
  } else if (item.page) {
    return item.page.id;
  } else if (item.url) {
    return item.url;
  } else {
    throw unknownTypeError;
  }
}

export function getDiff(
  originalTree: TreeItem[],
  newTree: TreeItem[],
): TreeOperation[] {
  const originalMap = treeToMap(originalTree, "root");
  const newMap = treeToMap(newTree, "root");

  const diff: TreeOperation[] = Object.keys(newMap).flatMap(key => {
    const originalNode = originalMap[key];
    const newNode = newMap[key];

    const patch = getPatch(originalNode, newNode);

    if (patch.length > 0) {
      const addedNode = patch.find(operation => operation.type === "add");
      const removedNode = patch.find(operation => operation.type === "remove");

      if (!!addedNode) {
        const changedParent = originalNode.length !== newNode.length;
        const sortOrder = removedNode
          ? addedNode.newPos - removedNode.oldPos
          : addedNode.newPos;

        // This exists because backend doesn't recongize the position of the new node
        // when it's moved from child to parent and/or up
        // We have to make an additional move so that backend can sort the new tree correctly
        // because without it the new node goes to the end of the parent node array by default
        // SimulatedMove is removed before submit
        if (changedParent && sortOrder !== originalNode.length) {
          return [
            {
              id: addedNode.items[0],
              parentId: key === "root" ? undefined : key,
              sortOrder: newNode.length - 1,
              type: "move" as TreeOperationType,
              simulatedMove: true,
            },
            {
              id: addedNode.items[0],
              parentId: key === "root" ? undefined : key,
              sortOrder:
                sortOrder - newNode.length < 0
                  ? sortOrder - newNode.length + 1
                  : sortOrder - newNode.length - 1,
              type: "move" as TreeOperationType,
            },
          ];
        }

        return {
          id: addedNode.items[0],
          parentId: key === "root" ? undefined : key,
          sortOrder,
          type: "move" as TreeOperationType,
        };
      }
    }
  });

  return diff.filter(d => !!d);
}

export function getNodeData(
  item: MenuDetailsFragment["items"][0],
  onChange: (operations: TreeOperation[]) => void,
  onClick: (id: string, type: MenuItemType) => void,
  onEdit: (id: string) => void,
): TreeItem {
  return {
    children: item.children.map(child =>
      getNodeData(child, onChange, onClick, onEdit),
    ),
    expanded: true,
    id: item.id,
    onChange,
    onClick: () => onClick(getItemId(item), getItemType(item)),
    onEdit: () => onEdit(item.id),
    title: item.name,
  };
}

export function getNodeQuantity(items: MenuDetailsFragment["items"]): number {
  return items.reduce(
    (acc, curr) => acc + getNodeQuantity(curr.children),
    items.length,
  );
}
