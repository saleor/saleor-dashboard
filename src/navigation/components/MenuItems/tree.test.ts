import {
  addNodeUnderParent,
  find,
  insertNode,
  removeNode,
} from "react-sortable-tree";

import { getDiff, MenuTreeItem } from "./tree";

const originalTree: MenuTreeItem[] = [
  {
    children: [
      { children: [], expanded: true, id: "0jewelry", title: "Jewelry" },
      { children: [], expanded: true, id: "1glasses", title: "Glasses" },
    ],
    expanded: true,
    id: "2accessories",
    title: "Accessories",
  },
  { children: [], expanded: true, id: "3groceries", title: "Groceries" },
  { children: [], expanded: true, id: "4apparel", title: "Apparel" },
];

function getNodeKey(node: any) {
  return node.treeIndex;
}

function moveNode(
  tree: MenuTreeItem[],
  src: string,
  target: string,
  asChild: boolean,
) {
  const { matches: srcNodeCandidates } = find({
    getNodeKey,
    searchMethod: ({ node }) => node.id === src,
    treeData: tree,
  });
  const srcNodeData = srcNodeCandidates[0];

  const treeAfterRemoval = removeNode({
    getNodeKey,
    path: srcNodeData.path,
    treeData: tree,
  }).treeData;

  const { matches: targetNodeCandidates } = find({
    getNodeKey,
    searchMethod: ({ node }) => node.id === target,
    treeData: treeAfterRemoval,
  });
  const targetNodeData = targetNodeCandidates[0];

  const treeAfterInsertion = asChild
    ? addNodeUnderParent({
        addAsFirstChild: true,
        getNodeKey,
        ignoreCollapsed: false,
        newNode: srcNodeData.node,
        parentKey: targetNodeData.treeIndex,
        treeData: treeAfterRemoval,
      }).treeData
    : insertNode({
        depth: targetNodeData.path.length,
        getNodeKey,
        minimumTreeIndex: targetNodeData.treeIndex,
        newNode: srcNodeData.node,
        treeData: treeAfterRemoval,
      }).treeData;

  return treeAfterInsertion as MenuTreeItem[];
}

describe("Properly computes diffs", () => {
  const testTable = [
    moveNode(originalTree, "1glasses", "0jewelry", true),
    moveNode(originalTree, "1glasses", "0jewelry", false),
    moveNode(originalTree, "2accessories", "4apparel", true),
  ];

  testTable.forEach(testData =>
    it("#", () => {
      const diff = getDiff(originalTree, testData);
      expect(diff).toMatchSnapshot();
    }),
  );
});
