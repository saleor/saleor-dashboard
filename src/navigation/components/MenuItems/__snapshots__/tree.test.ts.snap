// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Properly computes diffs # 1`] = `
Array [
  Object {
    "id": "1glasses",
    "parentId": "0jewelry",
    "sortOrder": 0,
    "type": "move",
  },
]
`;

exports[`Properly computes diffs # 2`] = `
Array [
  Object {
    "id": "1glasses",
    "parentId": "2accessories",
    "sortOrder": -1,
    "type": "move",
  },
]
`;

exports[`Properly computes diffs # 3`] = `
Array [
  Object {
    "id": "2accessories",
    "parentId": "4apparel",
    "sortOrder": 0,
    "type": "move",
  },
]
`;
